### DISCUSSION TITLE
Get field and refine errors from zodValidator at the same time

### DISCUSSION THREAD

**Discussion by elieven** [react-hook-form/react-hook-form#10068] [Q&A] [answered]
I'm trying to make a form that sets the % of the total payout to be divided between people. To do this I'm using the following zod schema:

```ts
const formDataSchema = z.object({
  payouts: z
    .array(
      z.object({
        percentage: z
          .number()
          .min(0, { message: "Percentage must not be less that 0" })
          .max(100, { message: "Percentage must not be more than 100" })
      })
    )
    .refine((val) => val.reduce((a, v) => a + v.percentage, 0) === 100, {
      message: "The sum of percentages must be exactly 100"
    })
});
```

My problem is that if there is a refine error (_in this case if the sum is not 100 exactly_) the refine error is the only one that i get n the `errors.payouts` object. There is no way to check if the fields themselves have a wrong input like for example -10 or 150.

I know there are some hacky half-ways to fix it like setting a min and max on the input (which i do) but the user can just type in whatever and also there's the "clamp the value on change" trick and all that but I'd love it if there was a way to just get the damn errors for everything not just the refinement one. 

Also once the refine error is gone and the sum is 100 only then do other errors appear. How could i get all the errors at all times?

Link to [codesandbox here](https://codesandbox.io/s/rhf-zod-validator-refine-error-bs-1hzrux).

**Comment by leapful** [comment, marked as answer]
@elieven 

By default, zod resolver from @hookform library will collect error by field path where `refine()` will stand for field path "payouts" in your example and overrides the whole error path of each item inside `payouts`.

Just define a custom field path for validation message from `refine()` to get rid of this issue.

Working demo: https://codesandbox.io/s/rhf-zod-validator-refine-error-bs-forked-r8riby?file=/src/App.tsx

**Reply by elieven** [reply]
Ok I think I understand, thanks a bunch. Will look into these error paths more closely. How come I can't see the `errors.payouts.refine` in the object logged to console and also ts is complaining that there is no such thing?

**Reply by leapful** [reply]
@elieven 

The whole array `payouts` will have error object as an array too, the custom path `refine` is added custom property to that array therefore it cannot be shown in `console.log` and type checking will not accept it. If you are ok with this quick workaround, just force cast type for `refine` to comfort type checking. Otherwise, you have to implement custom resolver to collect errors manually after validating with zod.

**Reply by elieven** [reply]
I thank you for all the help thus far and I'd hate to burden you any further but I'd love it if you could provide an example as it might be helpful for anyone with a similar question in the future (_it seems to me it might be relatively common if you go one layer deeper than just the standard form_) and since I wasn't able to find anything in the docs as far as explanations and examples for this sort of thing goes.
